### DISCUSSION TITLE
Passing server_config to APIHandler inside add_routes

### DISCUSSION THREAD

**Discussion by nat-n** [langchain-ai/langserve#459] [Ideas]
I'm using `add_routes` and would like to set pass a `RunnableConfig` as `server_config` to the chain that depends on the contents of the request object (such as request headers). 

The only way I see to do this today is by using `APIHandler` directly, but it would be nice to still benefit from `add_routes`. One solution would be to make `add_routes` accept an additional optional parameter like:

```python
server_config_mapper: Optional[Callable[[Request], RunnableConfig]] = None
```

Which is then used in each relevant route handler like so:

```python
        async def invoke(request: Request) -> Response:
            server_config = (
                server_config_mapper(request) if server_config_mapper else None
            )

            return await api_handler.invoke(request, server_config=server_config)
```

I've implemented this with a local fork of the server.py module and would be happy to contribute a PR if the maintainers agree that it's a good solution?

Another approach would be to decompose `add_routes` into smaller reusable units, so a user can create their own `APIHandler` object (potentially with custom extensions such as decorating in invoke method to derive server_config from the request) and recompose the behavior that want around it. This would be more powerful but require a larger refactor to expand the API surface area.

Is there another way to approach this that I've missed?

For more context, my use case is that I want to add metadata to LangSmith traces that depends on the request headers, but one could easily image wanting to set other RunnableConfig attributes in the same way to control chain behavior.

**Comment by eyurtsev** [comment]
Did you take a look at the per request modifier param in add routes? It allows modifying the config based on the request

**Reply by nat-n** [reply]
Thanks for the reply @eyurtsev. I totally just didn't see that there... and it seems to do exactly what I need. Nice!

**Comment by eyurtsev** [comment]
And thank you for writing Poe the poet!

**Comment by eyurtsev** [comment]
Closing as there's currently a way to do this
